### DISCUSSION TITLE
Better manage Zim modules and prompts allowing the project to grow in an organized way

### DISCUSSION THREAD

**Discussion by ericbn** [zimfw/zimfw#88] [Ideas]
I believe Oh My Zsh and even Prezto failed to grow in an organized fashion mainly because they lack incorporating some kind of plugin system that would allow people to contribute with external modules and keep the core clean and simple.

I suggest we take this step with Zim (and make it awesome)!

My suggestion would go in the line of:
- Creating a `zim` GitHub organization, having the Zim core be `zim/zim`, and other Zim modules be like `zim/xyz`.
- Keeping into the Zim core only what we would agree are the basic modules.
- Allowing users to configure external prompts as easily as having this in `.zimrc`:
  
  ```
  zprompt_theme='subnixr/minimal'
  ```
- Allowing users to declare external modules also as easily as (two last ones):
  
  ```
  zmodules=(directory environment git history input utility meta custom \
            syntax-highlighting history-substring-search prompt completion \
            zim/pacman foo/bar)
  ```

Maybe we can incorporate an external existing plugin system into the Zim core to accomplish this, or we can implement our own system (which I think would be the more flexible way to go).

The plugin system would have at least functions like:
- `install` to install new declared modules
- `update` to update all existing modules
- `clean`/`cleanup` to remove unused modules that are still installed

In case we use an internal plugin system, that could be managed by `zmanage`, which we already have. (Side suggestion: wouldn't it be cooler to rename `zmanage` to `zim`?)

Maybe the `zim` organization could be the default for external modules, so if the user declares simply `pacman` inside `zmodules` and there's no such module, the plugin system would try to check for a `zim/pacman` repo to install from there.

I would not bother making the plugin system capable of importing Oh My Zsh or Prezto modules.

My suggestion of the basic modules for the Zim code are:
- All those importing the Zsh ones: `completion`, `history-substring-search`, `syntax-highlighting`
- All the ones configuring the sane Zsh defaults for all/most OSs: `directory`, `environment`, `history`, `input`, `spectrum`(?), `ssh` (only not useful for Mac OS X), `utility`
- The zsh core ones: `debug` and `meta`
- `git`, as git is needed to install and keep Zim and external modules updated
- `prompt` with a small selection of internal prompts (we have `eriner`, `gitster`, `magicmace`, `steeef`; there's already a newer fork of `minimal` as reported in https://github.com/Eriner/zim/issues/78 and we could check if that is the case with any of the others)

I'd suggest `archive`(?), `fasd` and `pacman` go as external modules inside the `zim` organization.

This would instantly make users use whatever external prompt they want, without thinking about creating a "Add xyz external prompt" pull request. Other module requests like https://github.com/Eriner/zim/issues/33 (`virtualenv`) or https://github.com/Eriner/zim/issues/82 (a module for prompts showing git info) would become candidates for external modules, and we even could decide whether we choose to adopt the module inside the `zim` organization or suggest the contributor to keep it as a personal project inside its own GitHub user account.

Thoughts?

**Comment by Eriner** [comment]
I've only read the first line of your proposal; this is something I realized quickly after starting. It would be ideal if all the modules had their own repository, and zim/zim would be strictly the "loader" and associated tasks. I'll read the rest of your proposal and respond at some point this weekend. This would be a fairly large undertaking and would require a fair amount of code refactoring, so I see this as a good long-term goal.

**Comment by Plloi** [comment]
To maintain theme I propose callnowia or cbp (conveyor belt planet) as names for a utility for managing modules.

EDIT: Or possibly Vort.

**Comment by Eriner** [comment]
To follow this up, I have created an organization and will start to populate that sometime next week. In the meantime, any suggestions or discussion around this topic belongs here.

@ericbn:

> Side suggestion: wouldn't it be cooler to rename `zmanage` to `zim`?

No, I don't want to clobber the zim-wiki binary name.

In terms of the module organization, all of the current modules would get their **own** repositories within the organization; zim itself will be _only_ the loader (basically `init.zsh`, save some helper scripts). This will make module management much easier, and will allow me to designate users as maintainers of specific modules. For example, I don't own a Mac, so if we wanted a Mac-specific module, I wouldn't be maintaining that. 

@Plloi:

> To maintain theme I propose callnowia or cbp (conveyor belt planet) as names for a utility for managing modules.
>  oh come now, if we need a meta tool to fit the theme, it's `GIR` (or possibly `SIR`).

Also, I will likely be making the move from submodules to subtrees in terms of zim-core management.

**Comment by ericbn** [comment]
@Eriner, cool! I'd be happy to help coding whatever is needed.

**Comment by Eriner** [comment]
Thanks @ericbn, I (and I'm sure many other users) appreciate that, as well as the work you've done already! :) :+1:

**Comment by Plloi** [comment]
Not sure i would trust GIR with something so important. Plus there is fun to be have with callnowia and cbp...

callnowia (order|return) for install/uninstall

or a megadoomer easter egg for cbp ...Megadoomer!

**Comment by ericbn** [comment]
Just googled those names you guys are mentioning and found this [series](http://zim.wikia.com/wiki/Invader_Zim_%28series%29) I've never heard of (maybe because I don't watch Nickelodeon)... Is "Zim" really inspired in this series, and you seriously plan naming things after this?

**Comment by Plloi** [comment]
Well the graphic in the readme is Zim's(or at least an Irken) head, inspiration acheived!

As for naming everything else well that's up to @Eriner, but it would make me happy because [I love this show](https://www.youtube.com/watch?v=wdMHgc2HzQU)

**Comment by Eriner** [comment]
I am going to migrate this issue to the new issue tracker, so follow the reference below if you're looking for the current discussion.

**Comment by Eriner** [comment]
This issue was moved to zimframework/zim#2
